src/strategies/masterchef-pool-balance-price/index.ts

```diff
@@ -0,0 +1,450 @@
+
+    printLog();
+
+    const tokenCount = poolStaked
+      .mul(tokensPerLp)
+      .div(tokenDecimals)
+      .mul(weight)
+      .div(weightDecimals);
+
+    log.push(`tokenCount = ${tokenCount}`);
+
+    return toFloat(tokenCount, _options.decimals) * price;
+}
+
+function printLog(){
+  if(_options.log || false)
+  {
+    console.debug(log);
+    log = []
+  }
+}
+
+async function getTokenPrice(tokenAddress: any, network: any, provider: any, blockTag: string | number) {
+  let price: number = 1;
+  const cacheKey = tokenAddress + blockTag;
+
+  if (_options.usePrice === true && !priceCache.has(cacheKey)) {
+     log.push(`calling getPrice for token address: ${tokenAddress} and blockTag: ${blockTag}`);
+
+     priceCache.set(cacheKey, await getPrice(network, provider, tokenAddress, blockTag) || 1);
+  }
+
+  price = priceCache.get(cacheKey) || 1;
+
+  return price;
+}
+
+async function getPrice(
+  network,
+  provider,
+  address,
+  blockTag) {
+  if(!blockCache.has(blockTag)){
+    blockCache.set(blockTag, await provider.getBlock(blockTag));
+  }
+
+  const block = blockCache.get(blockTag);
+  const platform = networksWithPlatforms[network];
+  const currency = _options.currency || 'usd'
+  const from = block.timestamp - 100000;
+  const to = block.timestamp;
+  const coingeckoApiURL = `https://api.coingecko.com/api/v3/coins/${platform}/contract/${new String(address).toLowerCase()}/market_chart/range?vs_currency=${currency}&from=${from}&to=${to}`;
+
+  log.push(`platform = ${platform}`);
+  log.push(`from = ${from}`);
+  log.push(`to = ${from}`);
+  log.push(`coingeckoApiURL = ${coingeckoApiURL}`);
+
+  const coingeckoData = await fetch(coingeckoApiURL)
+    .then(async (r) => {
+      log.push(`coingeco response = ${inspect(r)}`);
+
+      const json = await r.json();
+
+      log.push(`coingecko json = ${inspect(json)}`);
+
+      return json;
+    })
+    .catch((e) => {
+      console.error(e);
+      throw new Error('Strategy masterchef-pool-balance-of-token: coingecko api failed');
+    });
+
+  return (coingeckoData.prices?.pop()?.pop() || 0) as number;
+}
+
+export async function strategy(
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
+) {
+  _options = options;
+  const blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
+  const userInfoCalls = getUserInfoCalls(addresses);
+  const tokenCalls = getTokenCalls();
+  const entries = new Map<PropertyKey,any>();
+
+
+  const userInfoResponse = await multicall(
+    network,
+    provider,
+    abi,
+    userInfoCalls,
+    { blockTag }
+  );
+
+  const userInfoChunks = arrayChunk(
+    userInfoResponse,
+    1
+  );
+
+  const tokenResponse = await multicall(
+    network,
+    provider,
+    abi,
+    tokenCalls,
+    { blockTag }
+  );
+
+  for (let i = 0; i < userInfoChunks.length; i++) {
+    const value = userInfoChunks[i];
+    const score = await processValues(value, tokenResponse, network, provider, blockTag)
```

Comment:
Behind this function i can see there is a lot of call and we can not predict the number of userInfoChunks, this might be too heavy to process, we usually recommend to limit a strategy with a maximum of 3 multicall or requests. Is there any way you can simplify the strategy so it does less calls?

Reply:
I made sure that scenario wouldn't happen I'm caching the price for each token on every execution of the strategy.
There are only 2 multicalls in total, behind the ProcessValues there are at most 2 requests to CoinGeco to fetch the prices. That call to CoinGecko is cached so there will be at the maximum only 2 requests. 
If the configuration is set to get the price for only 1 token, then the total will be 1 request.
If the configuration is not set to get the price then no request will be made.
If I did the single multicall like in the original masterchef-pool-balance, the test would fail at 400 addresses because of the size of the body.

Is 4 multicals or requests, a difference of 1, that impactful?

From the tests check it looked like the cache was working well.

Reply:
Got it, thanks for the explanation